### PR TITLE
swev-id: matplotlib__matplotlib-26113 — Align hexbin mincnt behavior with and without C (>= mincnt)

### LINKED ISSUE #63: Inconsistent behavior of hexbin mincnt parameter when C is provided (parity fix)
Inconsistent behavior of hexbins mincnt parameter, depending on C parameter

Task reference: ID 282

### Bug report

**Bug summary**

Different behavior of `hexbin`'s `mincnt` parameter, depending on whether the `C` parameter is supplied.

**Code for reproduction**

```python
from matplotlib import pyplot
import numpy as np

np.random.seed(42)

X, Y = np.random.multivariate_normal([0.0, 0.0], [[1.0, 0.1], [0.1, 1.0]], size=250).T
Z = np.ones_like(X)

extent = [-3., 3., -3., 3.]
gridsize = (7, 7)

# no mincnt, no C
fig, ax = pyplot.subplots(1, 1)
ax.hexbin(X, Y, extent=extent, gridsize=gridsize, linewidth=0.0, cmap='Blues')
ax.set_facecolor("green")

# mincnt=1, no C
fig, ax = pyplot.subplots(1, 1)
ax.hexbin(X, Y, mincnt=1, extent=extent, gridsize=gridsize, linewidth=0.0, cmap='Blues')
ax.set_facecolor("green")

# no mincnt, with C
fig, ax = pyplot.subplots(1, 1)
ax.hexbin(X, Y, C=Z, reduce_C_function=np.sum, extent=extent, gridsize=gridsize, linewidth=0.0, cmap='Blues')
ax.set_facecolor("green")

# mincnt=1, with C
fig, ax = pyplot.subplots(1, 1)
ax.hexbin(X, Y, C=Z, reduce_C_function=np.sum, mincnt=1, extent=extent, gridsize=gridsize, linewidth=0.0, cmap='Blues')
ax.set_facecolor("green")

# mincnt=0, with C
fig, ax = pyplot.subplots(1, 1)
ax.hexbin(X, Y, C=Z, reduce_C_function=np.sum, mincnt=0, extent=extent, gridsize=gridsize, linewidth=0.0, cmap='Blues')
ax.set_facecolor("green")
```

**Actual outcome**

- With no `C` specified, `mincnt=1` shows bins with at least one datum.
- With `C` specified but not `mincnt`, bins with at least one datum are shown.
- With `C` specified and `mincnt=1`, bins require at least two datapoints (unexpected off-by-one).

**Expected outcome**

With `mincnt == 1`, the same gridpoints should be plotted whether `C` is supplied or not—i.e., bins with at least one point should be included.

**Additional resources**

- Prior behavior discussion: https://github.com/matplotlib/matplotlib/commit/5b127df288e0ec91bc897c320c7399fc9c632ddd

### Research and proposed resolution (by Emerson Gray)

Code locations (branch `matplotlib__matplotlib-26113`):
- File: `lib/matplotlib/axes/_axes.py`, `Axes.hexbin`
- C is None path effectively applies `counts >= mincnt` by setting `accum[accum < mincnt] = np.nan`.
- C is not None path uses `len(acc) > mincnt` to decide reducing via `reduce_C_function`, and defaults `mincnt` to `0` when None—leading to off-by-one relative to the C is None path.

Proposed fix:
- In the `C` is not None path:
  - If `mincnt is None`, set `mincnt = 1` (default excludes empty bins consistently).
  - Change reduction decision to `len(acc) >= mincnt` (parity with the C is None path).
- Update docstring for `mincnt` to state “at least mincnt” points instead of “more than mincnt”.

Potential side effects considered:
- Avoids calling `reduce_C_function([])` by defaulting `mincnt` to 1; parity achieved without breaking expected behavior.
- Masked arrays and color mapping continue to treat failing bins as NaN; unaffected.
- `bins='log'`/norm autoscale unaffected; marginals compute reduces where `len(ci) > 0` independent of mincnt.

Unit tests to add:
- Deterministic dataset with two points in separate bins: `X=Y=[0.1, 1.6]`, `gridsize=(2,2)`, `extent=(0,2,0,2)`.
  - Test A: `C=None`, `mincnt=1` includes both bins, values equal to 1.
  - Test B: `C=np.ones_like(X)`, `reduce_C_function=np.sum`, `mincnt=1` includes both bins, sums equal to 1.
  - Optional: `mincnt=2` excludes bins with a single point for both paths.

We will submit a PR implementing this change and tests, targeting base branch `matplotlib__matplotlib-26113`.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- set the `C`-aware hexbin path to treat `mincnt=None` as 1 and require `len(acc) >= mincnt`
- update the `mincnt` docstring to describe the \"at least\" threshold
- add focused regression tests covering parity between count- and `C`-based binning

## Issue
Fixes #63

## Reproduction
```python
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
x = np.array([0.1, 1.6])
y = np.array([0.1, 1.6])

print(ax.hexbin(x, y, gridsize=(2, 2), extent=(0, 2, 0, 2), mincnt=1).get_array())
print(ax.hexbin(x, y, C=np.ones_like(x), gridsize=(2, 2), extent=(0, 2, 0, 2), mincnt=1, reduce_C_function=np.sum).get_array())
```

## Observed Behavior
- Without `C`, the bin counts were returned when population size was `>= mincnt`.
- With `C`, bins of size exactly equal to `mincnt` were dropped because the reducer only ran on `len(acc) > mincnt`.

## Expected Behavior
- Both code paths should emit bins whenever the population length meets or exceeds `mincnt`.

## Testing
- `MPLBACKEND=Agg PYTHONPATH=/workspace/matplotlib/lib LD_LIBRARY_PATH=/root/.nix-profile/lib .venv/bin/pytest -q lib/matplotlib/tests/test_axes.py -k hexbin`
- `MPLBACKEND=Agg PYTHONPATH=/workspace/matplotlib/lib LD_LIBRARY_PATH=/root/.nix-profile/lib .venv/bin/flake8 lib/matplotlib/axes/_axes.py lib/matplotlib/tests/test_axes.py --extend-ignore=E721`

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Test & Lint Summary
- ..........                                                               [100%]
10 passed, 871 deselected in 1.03s → 10 passed, 871 deselected
-  → no lint errors (E721 ignored due to existing file-level violations)

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -5012,9 +5012,9 @@ def reduce_C_function(C: array) -> float
                 else:
                     Cs_at_i2[i2[i]].append(C[i])
             if mincnt is None:
-                mincnt = 0
+                mincnt = 1
             accum = np.array(
-                [reduce_C_function(acc) if len(acc) > mincnt else np.nan
+                [reduce_C_function(acc) if len(acc) >= mincnt else np.nan
```
Comment (noa-lucent) [inline lib/matplotlib/axes/_axes.py]: [major] This change now invokes  even when a bin has zero hits and  is explicitly set to 0. Previously we skipped those bins ( evaluated to ), so reducers never saw an empty list. With the new guard (),  would evaluate the reducer on , which regresses existing callers: the default  emits , and many reducers (, custom callables) raise outright. Please keep the protection against empty accumulators (e.g. require  in addition to the  threshold) so we never call reducers on impossible states.

Comment (noa-lucent) [inline lib/matplotlib/axes/_axes.py]: [major] With this guard loosened we now call `reduce_C_function` even when a bin has zero hits and the caller set `mincnt=0`. Previously we skipped those cases because `len(acc) > mincnt` evaluated to `False`, so reducers never saw an empty list. Invoking reducers on `[]` regresses existing users: the default `np.mean` emits `RuntimeWarning: Mean of empty slice`, and many reducers (`np.min`, custom callables) raise outright. We still need to avoid ever invoking the reducer on an empty accumulator—e.g. require `len(acc) > 0` in addition to the `mincnt` threshold—so this branch only runs when we have at least one value to reduce.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for the updates—this gets the parity story close. I did spot one regression: with `mincnt=0` we now run the reducer on empty accumulators, which warns or raises for many reducers (including the default `np.mean`). Could you restore a guard so we never invoke `reduce_C_function` on an empty list?

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (noa-lucent) [reply lib/matplotlib/axes/_axes.py]: [major] Apologies for the garbled formatting above—the underlying issue is the same as in my other comment: with the new `len(acc) >= mincnt` check, passing `mincnt=0` will run `reduce_C_function([])` for empty bins. Please restore a guard so reducers never see an empty list (e.g. also require `len(acc) > 0`).

**Comment by casey-brooks** [comment]
### Test & Lint Summary (update)
- ...........                                                              [100%]
11 passed, 871 deselected in 1.03s → 11 passed, 871 deselected
-  → no lint errors (E721 ignored due to existing file-level violations)

File: lib/matplotlib/axes/_axes.py (hunk 1 above)
Comment (noa-lucent) [reply lib/matplotlib/axes/_axes.py]: Thanks for adding the explicit threshold—`len(acc) >= max(mincnt, 1)` keeps the reducer away from empty accumulators while preserving the >= `mincnt` behavior for real bins. The new regression test around `mincnt=0` also covers the edge case nicely. I'm satisfied with this fix.

**Review by noa-lucent — APPROVED** [review]
The guard now keeps `reduce_C_function` away from empty accumulators, and the new regression test locks the `mincnt=0` parity in place. Looks good to me—thanks for the quick follow-up.
